### PR TITLE
HPCC-18189 Replace some SpinBlock instances with atomics

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith and @mckellyln please review

File: thorlcr/slave/slave.cpp [hunk 1]
```diff
@@ -52,7 +52,6 @@ MODULE_INIT(INIT_PRIORITY_STANDARD)
 ProcessSlaveActivity::ProcessSlaveActivity(CGraphElementBase *container) : CSlaveActivity(container), threaded("ProcessSlaveActivity", this)
 {
     processed = 0;
```
Comment (jakesmith) [inline thorlcr/slave/slave.cpp]: picky: I'd have moved this to also be initialized at declaration time now lastCycles is.

File: thorlcr/slave/slave.cpp [hunk 2]
```diff
@@ -83,16 +82,13 @@ void ProcessSlaveActivity::threadmain()
 #ifdef TIME_ACTIVITIES
         if (timeActivities)
         {
-            {
-                SpinBlock b(cycleLock);
-                lastCycles = get_cycles_now(); // serializeStats will reset
-            }
+            lastCycles.store(get_cycles_now()); // serializeStats will reset
```
Comment (jakesmith) [inline thorlcr/slave/slave.cpp]: any reason this isn't lastCycles = get_cycles_now(); ?

**Comment by jakesmith** [comment]
@ghalliday - looks good, 2 tiny comments.
@mckellyln - tagging to review also.

File: thorlcr/slave/slave.cpp (hunk 2 above)
Comment (ghalliday) [reply thorlcr/slave/slave.cpp]: no - only to emphasise it is an atomic operation paired with the exchange below.  I can change if you prefer.

Comment (jakesmith) [reply thorlcr/slave/slave.cpp]: not that fussed, but seems a little inconsistent if done here and not elsewhere.
Personally I'd make it lastCycles = get_cycles_now();

File: thorlcr/slave/slave.cpp [hunk 3]
```diff
@@ -173,12 +168,13 @@ void ProcessSlaveActivity::serializeStats(MemoryBuffer &mb)
...
-        SpinBlock b(cycleLock);
-        if (lastCycles)
+        unsigned __int64 curCycles = lastCycles;
+        if (curCycles)
         {
             unsigned __int64 nowCycles = get_cycles_now();
-            totalCycles.totalCycles += nowCycles-lastCycles;
-            lastCycles = nowCycles; // time accounted for
+            //Update lastCycles to the current number of cycles - unless it has been set to 0 in the meantime
+            if (lastCycles.compare_exchange_strong(curCycles, nowCycles))
```
Comment (mckellyln) [inline thorlcr/slave/slave.cpp]: Do we really need seq_cst memory order ?

**Comment by mckellyln** [comment]
@ghalliday looks good.  I do ask about memory order - seq_cst is fine but is it required ?

File: thorlcr/slave/slave.cpp (hunk 3 above)
Comment (ghalliday) [reply thorlcr/slave/slave.cpp]: no, I will reduce it.

**Comment by ghalliday** [comment]
@mckellyln changed the cas(), and squashed the commits.

**Comment by mckellyln** [comment]
@ghalliday looks good to merge

**Review by mckellyln — APPROVED** [review]

**Comment by ghalliday** [comment]
@richardkchapman please sanity check and merge if happy

**Pull request merged**